File: bingo/expressions/agraph/simplification/cas_expression.py
"""Mathematical expression in tree form for CAS simplification.

This is the internal tree representation used by the simplification
pipeline.  It is *not* the same as :class:`AGraphExpression` — it is a
recursive tree of operator nodes used purely for algebraic manipulation.
"""

from ..operators import (
    POWER,
    INTEGER,
    MULTIPLICATION,
    CONSTANT,
    VARIABLE,
    ADDITION,
    SUBTRACTION,
    DIVISION,
    SAFE_POWER,
    SQUARE,
    CUBE,
    SQRT,
    ABS,
    EXPONENTIAL,
    LOGARITHM,
    SIN,
    COS,
    TAN,
    ARCSIN,
    ARCCOS,
    ARCTAN,
    SINH,
    COSH,
    TANH,
)

# Sentinel used to distinguish "not yet computed" from a cached ``None``
# value (which is a valid result for INTEGER operands).
_SENTINEL = object()

# ------------------------------------------------------------------ #
#  Interned singletons — allocated once, reused everywhere.           #
# ------------------------------------------------------------------ #
#
# Initialised lazily by _init_singletons() on first access so that
# CASExpression is defined before we try to instantiate them.

_ZERO = None
_ONE = None
_TWO = None
_NEG_ONE = None
_VARIABLE_CACHE = None  # list[CASExpression] for VARIABLE(0)..VARIABLE(7)

# ------------------------------------------------------------------ #
#  Operator sort-key — matches the old symbolic_regression ordering   #
#  so that commutative operand placement is identical.                #
# ------------------------------------------------------------------ #

_OPERATOR_ORDER = {
    INTEGER: -1,
    VARIABLE: 0,
    CONSTANT: 1,
    ADDITION: 2,
    SUBTRACTION: 3,
    MULTIPLICATION: 4,
    DIVISION: 5,
    SIN: 6,
    COS: 7,
    EXPONENTIAL: 8,
    LOGARITHM: 9,
    POWER: 10,
    ABS: 11,
    SQRT: 12,
    SAFE_POWER: 13,
    SINH: 14,
    COSH: 15,
    TAN: 16,
    ARCSIN: 17,
    ARCCOS: 18,
    ARCTAN: 19,
    TANH: 20,
    SQUARE: 21,
    CUBE: 22,
}


def _init_singletons():
    """Bootstrap module-level interned CASExpression singletons."""
    global _ZERO, _ONE, _TWO, _NEG_ONE, _VARIABLE_CACHE  # noqa: PLW0603
    if _ONE is not None:
        return
    _ZERO = CASExpression(INTEGER, [0])
    _ONE = CASExpression(INTEGER, [1])
    _TWO = CASExpression(INTEGER, [2])
    _NEG_ONE = CASExpression(INTEGER, [-1])
    _VARIABLE_CACHE = [CASExpression(VARIABLE, [i]) for i in range(8)]


def get_interned_integer(value):
    """Return a cached CASExpression for common integer values."""
    _init_singletons()
    if value == 0:
        return _ZERO
    if value == 1:
        return _ONE
    if value == 2:
        return _TWO
    if value == -1:
        return _NEG_ONE
    return CASExpression(INTEGER, [value])


def get_interned_variable(index):
    """Return a cached CASExpression for common variable indices."""
    _init_singletons()
    if 0 <= index < len(_VARIABLE_CACHE):
        return _VARIABLE_CACHE[index]
    return CASExpression(VARIABLE, [index])


def _get_one():
    _init_singletons()
    return _ONE


class CASExpression:
    """A mathematical expression in tree form.

    Parameters
    ----------
    operator : int
        Operator ID from :mod:`..operators`.
    operands : list
        Operands — each is either a :class:`CASExpression` or an ``int``
        (for terminal parameter values).
    """

    __slots__ = (
        "_operator",
        "_operands",
        "_is_constant_valued",
        "_depends_on",
        "_hash",
        "_base",
        "_exponent",
        "_term",
        "_coefficient",
    )

    def __init__(self, operator, operands):
        self._operator = operator
        self._operands = operands
        self._is_constant_valued = None
        self._depends_on = None
        self._hash = None
        self._base = _SENTINEL
        self._exponent = _SENTINEL
        self._term = _SENTINEL
        self._coefficient = _SENTINEL

    @property
    def operator(self):
        """The primary mathematical operation in the expression."""
        return self._operator

    @property
    def operands(self):
        """The operands upon which the expression's operator acts."""
        return self._operands

    @property
    def is_constant_valued(self):
        """Whether the expression is derived solely from constant values."""
        if self._is_constant_valued is None:
            self._is_constant_valued = self._is_derived_from_constants()
        return self._is_constant_valued

    @property
    def depends_on(self):
        """The constants, integers and variables in sub-expressions."""
        if self._depends_on is None:
            self._depends_on = self._find_what_expression_depends_on()
        return self._depends_on

    @property
    def base(self):
        """The base *x* in *x^b*."""
        if self._base is _SENTINEL:
            if self._operator == POWER:
                self._base = self._operands[0]
            elif self._operator == INTEGER:
                self._base = None
            else:
                self._base = self
        return self._base

    @property
    def exponent(self):
        """The exponent *b* in *x^b*."""
        if self._exponent is _SENTINEL:
            if self._operator == POWER:
                self._exponent = self._operands[1]
            elif self._operator == INTEGER:
                self._exponent = None
            else:
                self._exponent = _get_one()
        return self._exponent

    @property
    def term(self):
        """The term *x* in *A·x*."""
        if self._term is _SENTINEL:
            if self._operator == MULTIPLICATION:
                if self._operands[0].operator in [INTEGER, CONSTANT]:
                    self._term = CASExpression(MULTIPLICATION, self._operands[1:])
                else:
                    self._term = self
            elif self._operator == INTEGER:
                self._term = None
            else:
                self._term = CASExpression(MULTIPLICATION, [self])
        return self._term

    @property
    def coefficient(self):
        """The coefficient *A* in *A·x*."""
        if self._coefficient is _SENTINEL:
            if self._operator == MULTIPLICATION and self._operands[0].operator in [
                INTEGER,
                CONSTANT,
            ]:
                self._coefficient = self._operands[0]
            elif self._operator == INTEGER:
                self._coefficient = None
            else:
                self._coefficient = _get_one()
        return self._coefficient

    # ------------------------------------------------------------------ #
    #  Private helpers                                                    #
    # ------------------------------------------------------------------ #

    def _is_derived_from_constants(self):
        if self._operator in [INTEGER, CONSTANT]:
            return True
        if self._operator == VARIABLE:
            return False
        for operand in self._operands:
            if not operand.is_constant_valued:
                return False
        return True

    def _find_what_expression_depends_on(self):
        if self._operator == INTEGER:
            return {"i"}
        if self._operator == VARIABLE:
            return {"x"}
        if self._operator == CONSTANT:
            return {self._operands[0]}
        return set.union(*[o.depends_on for o in self._operands])

    # ------------------------------------------------------------------ #
    #  Public helpers                                                     #
    # ------------------------------------------------------------------ #

    def map(self, mapped_function):
        """Apply a function to all operands of the expression.

        Returns ``self`` when no operand changed (identity short-circuit).

        Parameters
        ----------
        mapped_function : callable

        Returns
        -------
        CASExpression
        """
        mapped_operands = [mapped_function(i) for i in self._operands]
        if all(m is o for m, o in zip(mapped_operands, self._operands)):
            return self
        return CASExpression(self._operator, mapped_operands)

    def is_zero(self):
        """Whether the expression is the integer ``0``."""
        if self._operator != INTEGER:
            return False
        return self._operands[0] == 0

    def is_one(self):
        """Whether the expression is the integer ``1``."""
        if self._operator != INTEGER:
            return False
        return self._operands[0] == 1

    def same_term(self, other):
        """Check if *self* and *other* share the same additive term.

        Equivalent to ``self.term == other.term`` but avoids allocating
        wrapper ``CASExpression(MULTIPLICATION, [x])`` nodes that the
        ``.term`` property creates for non-MULTIPLICATION operands.

        Returns ``False`` for INTEGER-typed operands (which have term=None).
        """
        s_op = self._operator
        o_op = other._operator

        if s_op == INTEGER or o_op == INTEGER:
            return False

        s_has_coeff = s_op == MULTIPLICATION and self._operands[0].operator in (
            INTEGER,
            CONSTANT,
        )
        o_has_coeff = o_op == MULTIPLICATION and other._operands[0].operator in (
            INTEGER,
            CONSTANT,
        )

        if s_has_coeff and o_has_coeff:
            return self._operands[1:] == other._operands[1:]
        if s_has_coeff and not o_has_coeff:
            s_tail = self._operands[1:]
            # other is also MULTIPLICATION (without leading coeff) →
            # compare tail operands directly.
            if o_op == MULTIPLICATION:
                return list(s_tail) == list(other._operands)
            return len(s_tail) == 1 and s_tail[0] == other
        if not s_has_coeff and o_has_coeff:
            o_tail = other._operands[1:]
            # self is also MULTIPLICATION (without leading coeff) →
            # compare operands directly.
            if s_op == MULTIPLICATION:
                return list(self._operands) == list(o_tail)
            return len(o_tail) == 1 and o_tail[0] == self
        # Neither has a leading coefficient — terms are
        # MULTIPLICATION([self]) vs MULTIPLICATION([other])
        return self == other

    def copy(self):
        """Shallow copy of the expression."""
        return CASExpression(self._operator, self._operands)

    # ------------------------------------------------------------------ #
    #  Comparison / hashing                                               #
    # ------------------------------------------------------------------ #

    def __eq__(self, other):
        if self is other:
            return True
        if other is None:
            return False
        if self._operator != other.operator:
            return False
        return self._operands == other.operands

    def __ne__(self, other):
        return not self.__eq__(other)

    def __lt__(self, other):
        if self.is_constant_valued or other.is_constant_valued:
            return self._constant_lt(other)

        s_op = self._operator
        o_op = other.operator
        if MULTIPLICATION in (s_op, o_op):
            return self._associative_lt(other, MULTIPLICATION)
        if POWER in (s_op, o_op):
            return self._power_lt(other)
        if ADDITION in (s_op, o_op):
            return self._associative_lt(other, ADDITION)
        return self._general_lt(other)

    def _constant_lt(self, other):
        if self.is_constant_valued != other.is_constant_valued:
            return self.is_constant_valued
        return self._general_lt(other)

    def _general_lt(self, other):
        if self._operator != other.operator:
            return _OPERATOR_ORDER[self._operator] < _OPERATOR_ORDER[other.operator]
        return self._operands_lt(self._operands, other.operands)

    @staticmethod
    def _operands_lt(s_operands, o_operands):
        for s_operand, o_operand in zip(reversed(s_operands), reversed(o_operands)):
            if s_operand != o_operand:
                return s_operand < o_operand
        return len(s_operands) < len(o_operands)

    def _associative_lt(self, other, associative_operator):
        if self._operator == associative_operator:
            if other.operator == associative_operator:
                return self._operands_lt(self._operands, other.operands)
            return self._operands_lt(self._operands, [other])
        return self._operands_lt([self], other.operands)

    def _power_lt(self, other):
        s_base = self.base
        s_exponent = self.exponent
        o_base = other.base
        o_exponent = other.exponent

        if s_base == o_base:
            return s_exponent < o_exponent
        return s_base < o_base

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        string = f"{self._operator}("
        for operand in self._operands:
            string += f"{operand}, "
        string += ")"
        return string

    def __hash__(self):
        if self._hash is None:
            h = hash(self._operator)
            for op in self._operands:
                h = h * 31 ^ hash(op)
            self._hash = h
        return self._hash


# ------------------------------------------------------------------ #
#  Eagerly initialise singletons now that CASExpression is defined.   #
# ------------------------------------------------------------------ #
_init_singletons()

File: bingo/expressions/agraph/cppagraph/src/cas_expression.cpp
/**
 * @file cas_expression.cpp
 * @brief CASExpression implementation — port of cas_expression.py.
 */

#include "cppagraph/cas_expression.h"

#include <algorithm>
#include <cassert>
#include <sstream>
#include <unordered_map>

namespace cppagraph {

static inline uint8_t u8(Op o) { return static_cast<uint8_t>(o); }

// ------------------------------------------------------------------ //
//  Operator sort-key table (matches Python _OPERATOR_ORDER)           //
// ------------------------------------------------------------------ //

static const std::unordered_map<uint8_t, int> OPERATOR_ORDER_MAP = {
    {u8(Op::INTEGER),        -1},
    {u8(Op::VARIABLE),        0},
    {u8(Op::CONSTANT),        1},
    {u8(Op::ADDITION),        2},
    {u8(Op::SUBTRACTION),     3},
    {u8(Op::MULTIPLICATION),  4},
    {u8(Op::DIVISION),        5},
    {u8(Op::SIN),             6},
    {u8(Op::COS),             7},
    {u8(Op::EXPONENTIAL),     8},
    {u8(Op::LOGARITHM),       9},
    {u8(Op::POWER),          10},
    {u8(Op::ABS),            11},
    {u8(Op::SQRT),           12},
    {u8(Op::SAFE_POWER),     13},
    {u8(Op::SINH),           14},
    {u8(Op::COSH),           15},
    {u8(Op::TAN),            16},
    {u8(Op::ARCSIN),         17},
    {u8(Op::ARCCOS),         18},
    {u8(Op::ARCTAN),         19},
    {u8(Op::TANH),           20},
    {u8(Op::SQUARE),         21},
    {u8(Op::CUBE),           22},
};

int operator_order(uint8_t op) {
    auto it = OPERATOR_ORDER_MAP.find(op);
    return it != OPERATOR_ORDER_MAP.end() ? it->second : 99;
}

// ------------------------------------------------------------------ //
//  Interned singletons                                                //
// ------------------------------------------------------------------ //

static CASExprPtr g_zero;
static CASExprPtr g_one;
static CASExprPtr g_two;
static CASExprPtr g_neg_one;
static std::vector<CASExprPtr> g_variable_cache;
static bool g_singletons_init = false;

static void init_singletons() {
    if (g_singletons_init) return;
    g_zero    = std::make_shared<CASExpression>(u8(Op::INTEGER), 0);
    g_one     = std::make_shared<CASExpression>(u8(Op::INTEGER), 1);
    g_two     = std::make_shared<CASExpression>(u8(Op::INTEGER), 2);
    g_neg_one = std::make_shared<CASExpression>(u8(Op::INTEGER), -1);
    g_variable_cache.reserve(8);
    for (int i = 0; i < 8; ++i)
        g_variable_cache.push_back(
            std::make_shared<CASExpression>(u8(Op::VARIABLE), i));
    g_singletons_init = true;
}

CASExprPtr cas_zero()    { init_singletons(); return g_zero; }
CASExprPtr cas_one()     { init_singletons(); return g_one; }
CASExprPtr cas_two()     { init_singletons(); return g_two; }
CASExprPtr cas_neg_one() { init_singletons(); return g_neg_one; }

CASExprPtr interned_integer(int value) {
    init_singletons();
    if (value == 0)  return g_zero;
    if (value == 1)  return g_one;
    if (value == 2)  return g_two;
    if (value == -1) return g_neg_one;
    return std::make_shared<CASExpression>(u8(Op::INTEGER), value);
}

CASExprPtr interned_variable(int index) {
    init_singletons();
    if (index >= 0 && index < static_cast<int>(g_variable_cache.size()))
        return g_variable_cache[index];
    return std::make_shared<CASExpression>(u8(Op::VARIABLE), index);
}

// ------------------------------------------------------------------ //
//  Construction                                                       //
// ------------------------------------------------------------------ //

CASExpression::CASExpression(uint8_t op, int param)
    : _op(op), _terminal_param(param) {}

CASExpression::CASExpression(uint8_t op, std::vector<CASExprPtr> operands)
    : _op(op), _terminal_param(0), _operands(std::move(operands)) {}

// ------------------------------------------------------------------ //
//  Predicates                                                         //
// ------------------------------------------------------------------ //

bool CASExpression::is_zero() const {
    return _op == u8(Op::INTEGER) && _terminal_param == 0;
}

bool CASExpression::is_one() const {
    return _op == u8(Op::INTEGER) && _terminal_param == 1;
}

bool CASExpression::is_constant_valued() {
    if (_is_constant_valued_cache == Tri::YES) return true;
    if (_is_constant_valued_cache == Tri::NO) return false;
    bool val = _compute_is_constant_valued();
    _is_constant_valued_cache = val ? Tri::YES : Tri::NO;
    return val;
}

bool CASExpression::_compute_is_constant_valued() {
    if (_op == u8(Op::INTEGER) || _op == u8(Op::CONSTANT)) return true;
    if (_op == u8(Op::VARIABLE)) return false;
    for (auto& child : _operands)
        if (!child->is_constant_valued()) return false;
    return true;
}

const CASExpression::DepSet& CASExpression::depends_on() {
    if (!_deps_computed) {
        _depends_on = _compute_depends_on();
        _deps_computed = true;
    }
    return _depends_on;
}

CASExpression::DepSet CASExpression::_compute_depends_on() {
    if (_op == u8(Op::INTEGER))  return {std::string("i")};
    if (_op == u8(Op::VARIABLE)) return {std::string("x")};
    if (_op == u8(Op::CONSTANT)) return {_terminal_param};
    DepSet result;
    for (auto& child : _operands) {
        auto& child_deps = child->depends_on();
        result.insert(child_deps.begin(), child_deps.end());
    }
    return result;
}

// ------------------------------------------------------------------ //
//  Algebraic properties                                               //
// ------------------------------------------------------------------ //

CASExprPtr CASExpression::base() {
    if (!_base_set) {
        _base_set = true;
        if (_op == u8(Op::POWER)) {
            _base = _operands[0];
        } else if (_op == u8(Op::INTEGER)) {
            _base = nullptr;
        } else {
            _base = shared_from_this();
        }
    }
    return _base;
}

CASExprPtr CASExpression::exponent() {
    if (!_exponent_set) {
        _exponent_set = true;
        if (_op == u8(Op::POWER)) {
            _exponent = _operands[1];
        } else if (_op == u8(Op::INTEGER)) {
            _exponent = nullptr;
        } else {
            _exponent = cas_one();
        }
    }
    return _exponent;
}

CASExprPtr CASExpression::term() {
    if (!_term_set) {
        _term_set = true;
        if (_op == u8(Op::MULTIPLICATION)) {
            auto& first = _operands[0];
            if (first->op() == u8(Op::INTEGER) || first->op() == u8(Op::CONSTANT)) {
                std::vector<CASExprPtr> tail(_operands.begin() + 1, _operands.end());
                _term = std::make_shared<CASExpression>(u8(Op::MULTIPLICATION), std::move(tail));
            } else {
                _term = shared_from_this();
            }
        } else if (_op == u8(Op::INTEGER)) {
            _term = nullptr;
        } else {
            _term = std::make_shared<CASExpression>(
                u8(Op::MULTIPLICATION),
                std::vector<CASExprPtr>{shared_from_this()});
        }
    }
    return _term;
}

CASExprPtr CASExpression::coefficient() {
    if (!_coefficient_set) {
        _coefficient_set = true;
        if (_op == u8(Op::MULTIPLICATION) &&
            (_operands[0]->op() == u8(Op::INTEGER) ||
             _operands[0]->op() == u8(Op::CONSTANT))) {
            _coefficient = _operands[0];
        } else if (_op == u8(Op::INTEGER)) {
            _coefficient = nullptr;
        } else {
            _coefficient = cas_one();
        }
    }
    return _coefficient;
}

// ------------------------------------------------------------------ //
//  Comparison / ordering                                              //
// ------------------------------------------------------------------ //

bool CASExpression::operator==(const CASExpression& other) const {
    if (this == &other) return true;
    if (_op != other._op) return false;
    if (IS_TERMINAL[_op]) return _terminal_param == other._terminal_param;
    if (_operands.size() != other._operands.size()) return false;
    for (size_t i = 0; i < _operands.size(); ++i)
        if (!(*_operands[i] == *other._operands[i])) return false;
    return true;
}

bool CASExpression::operator<(const CASExpression& other) const {
    // Need non-const for is_constant_valued; use const_cast (safe
    // since is_constant_valued only mutates cache).
    auto& self_nc = const_cast<CASExpression&>(*this);
    auto& other_nc = const_cast<CASExpression&>(other);
    if (self_nc.is_constant_valued() || other_nc.is_constant_valued())
        return _constant_lt(other);
    uint8_t s_op = _op, o_op = other._op;
    if (s_op == u8(Op::MULTIPLICATION) || o_op == u8(Op::MULTIPLICATION))
        return _associative_lt(other, u8(Op::MULTIPLICATION));
    if (s_op == u8(Op::POWER) || o_op == u8(Op::POWER))
        return _power_lt(other);
    if (s_op == u8(Op::ADDITION) || o_op == u8(Op::ADDITION))
        return _associative_lt(other, u8(Op::ADDITION));
    return _general_lt(other);
}

bool CASExpression::_constant_lt(const CASExpression& other) const {
    auto& self_nc = const_cast<CASExpression&>(*this);
    auto& other_nc = const_cast<CASExpression&>(other);
    if (self_nc.is_constant_valued() != other_nc.is_constant_valued())
        return self_nc.is_constant_valued();
    return _general_lt(other);
}

bool CASExpression::_general_lt(const CASExpression& other) const {
    if (_op != other._op)
        return operator_order(_op) < operator_order(other._op);
    if (IS_TERMINAL[_op])
        return _terminal_param < other._terminal_param;
    return _operands_lt(_operands, other._operands);
}

bool CASExpression::_operands_lt(const std::vector<CASExprPtr>& a,
                                  const std::vector<CASExprPtr>& b) {
    // Compare from the end (reversed iteration, matching Python)
    auto ai = a.rbegin(), bi = b.rbegin();
    for (; ai != a.rend() && bi != b.rend(); ++ai, ++bi) {
        if (!(**ai == **bi)) return **ai < **bi;
    }
    return a.size() < b.size();
}

bool CASExpression::_associative_lt(const CASExpression& other,
                                     uint8_t assoc_op) const {
    if (_op == assoc_op) {
        if (other._op == assoc_op)
            return _operands_lt(_operands, other._operands);
        return _operands_lt(_operands, {const_cast<CASExpression&>(other).shared_from_this()});
    }
    return _operands_lt(
        {const_cast<CASExpression&>(*this).shared_from_this()},
        other._operands);
}

bool CASExpression::_power_lt(const CASExpression& other) const {
    auto s_base = const_cast<CASExpression*>(this)->base();
    auto s_exp  = const_cast<CASExpression*>(this)->exponent();
    auto o_base = const_cast<CASExpression*>(&other)->base();
    auto o_exp  = const_cast<CASExpression*>(&other)->exponent();
    if (*s_base == *o_base)
        return *s_exp < *o_exp;
    return *s_base < *o_base;
}

// ------------------------------------------------------------------ //
//  same_term                                                          //
// ------------------------------------------------------------------ //

bool CASExpression::same_term(const CASExpression& other) {
    if (_op == u8(Op::INTEGER) || other._op == u8(Op::INTEGER))
        return false;

    bool s_has_coeff = (_op == u8(Op::MULTIPLICATION) &&
                        (_operands[0]->op() == u8(Op::INTEGER) ||
                         _operands[0]->op() == u8(Op::CONSTANT)));
    bool o_has_coeff = (other._op == u8(Op::MULTIPLICATION) &&
                        (other._operands[0]->op() == u8(Op::INTEGER) ||
                         other._operands[0]->op() == u8(Op::CONSTANT)));

    auto tail_eq = [](const std::vector<CASExprPtr>& a, size_t a_start,
                      const std::vector<CASExprPtr>& b, size_t b_start) {
        if (a.size() - a_start != b.size() - b_start) return false;
        for (size_t i = 0; i < a.size() - a_start; ++i)
            if (!(*a[a_start + i] == *b[b_start + i])) return false;
        return true;
    };

    if (s_has_coeff && o_has_coeff)
        return tail_eq(_operands, 1, other._operands, 1);

    if (s_has_coeff && !o_has_coeff) {
        if (other._op == u8(Op::MULTIPLICATION))
            return tail_eq(_operands, 1, other._operands, 0);
        return _operands.size() == 2 && *_operands[1] == other;
    }

    if (!s_has_coeff && o_has_coeff) {
        if (_op == u8(Op::MULTIPLICATION))
            return tail_eq(_operands, 0, other._operands, 1);
        return other._operands.size() == 2 && *other._operands[1] == *this;
    }

    return *this == other;
}

// ------------------------------------------------------------------ //
//  Hash                                                               //
// ------------------------------------------------------------------ //

std::size_t CASExpression::hash() const {
    if (_hash_computed) return _hash_value;
    std::size_t h = std::hash<uint8_t>{}(_op);
    if (IS_TERMINAL[_op]) {
        h = h * 31 ^ std::hash<int>{}(_terminal_param);
    } else {
        for (auto& child : _operands)
            h = h * 31 ^ child->hash();
    }
    _hash_value = h;
    _hash_computed = true;
    return h;
}

// ------------------------------------------------------------------ //
//  Utility                                                            //
// ------------------------------------------------------------------ //

CASExprPtr CASExpression::map(std::function<CASExprPtr(const CASExprPtr&)> fn) const {
    if (IS_TERMINAL[_op])
        return const_cast<CASExpression*>(this)->shared_from_this();
    std::vector<CASExprPtr> mapped;
    mapped.reserve(_operands.size());
    bool changed = false;
    for (auto& child : _operands) {
        auto m = fn(child);
        if (m.get() != child.get()) changed = true;
        mapped.push_back(std::move(m));
    }
    if (!changed) return const_cast<CASExpression*>(this)->shared_from_this();
    return std::make_shared<CASExpression>(_op, std::move(mapped));
}

CASExprPtr CASExpression::copy() const {
    if (IS_TERMINAL[_op])
        return std::make_shared<CASExpression>(_op, _terminal_param);
    return std::make_shared<CASExpression>(_op, _operands);
}

std::string CASExpression::to_string() const {
    std::ostringstream ss;
    ss << int(_op) << "(";
    if (IS_TERMINAL[_op]) {
        ss << _terminal_param;
    } else {
        for (size_t i = 0; i < _operands.size(); ++i) {
            if (i > 0) ss << ", ";
            ss << _operands[i]->to_string();
        }
    }
    ss << ")";
    return ss.str();
}

}  // namespace cppagraph
